fix(core/dfn-panel): make dfn panel keyboard accessible

marcoscaceres opened the pull request:
closes #2825 

cc @jnurthen. Would appreciate your input if you have time.

jnurthen commented on src/core/dfn-panel.runtime.js:
```diff
@@ -97,10 +140,67 @@ function displayPanel(dfn, panel, { x, y }) {
+
+  // If we are displaying because of a keydown,
+  // we want to trap focus
+  if (eventType === "keydown") {
```
This doesn't work for screen readers. Screen readers fire a synthetic click event when activating links so none of this logic will work. You can't branch on event type for this logic. Why not move focus for a mouse user too? You can dismiss the dialog when they click anywhere else on the screen.

jnurthen reviewed (changes requested):
appears to work nicely for keyboard users but fails for screen reader users. I think this is due to assuming that screen readers fire key events (which they normally don't) when activating links.

marcoscaceres commented:
@jnurthen any screen reader in particular? It works really nicely with VoiceOver at least.

jnurthen commented:
@marcoscaceres - not for me.... also tried with NVDA and didn't work there either.

marcoscaceres commented:
Had a chat with @jnurthen and he schooled me on a whole bunch of things. Updated code to reflect what we talked about... obvious on reflection, but differentiating between key and click events was obviously a bad thing. Now fixed.

marcoscaceres commented:
I did some investigation into testing keyboard interaction via Karma and Jasmine.... unfortunately, I couldn't find any useful resources. I did some experiments locally, but it turned into a giant hack. 

We might need to eventually move to a framework that supports keyboard events. 

Having said that, the clicks() still work fine, and give us similar enough expected behavior, so I think we are ok with what we have when it comes to tests.

sidvishnoi commented on src/core/dfn.js:
```diff
@@ -17,6 +17,12 @@ export function run() {
     }
 
+    // Make it possible to reach dfns by tabbing, allowing keyboard actions as needed.
+    dfn.tabIndex = 0;
```
Let's add tabIndex and title etc. in dfn-panel.js. Will help if we want to add an option to disable the feature in future, and keeps related code together.

marcoscaceres replied:
Agree about moving the title attribute, but I'm conflicted about tabindex. I see tabbing the dfn's as independent of the dfn panel. The dfn-panel just happens to respond to the keyboard activation of this element type.

sidvishnoi replied:
I don't think there are any other reasons why `<dfn>` should be focusable.

marcoscaceres replied:
I'd still like to keep significant modifications to dfn elements in dfn.js. Feels more natural having it here.

marcoscaceres replied:
Ah, forgot that Indexed terms also have popup menu (and they are `<span>`!)... so relocating this code to dfn-panel.

marcoscaceres replied:
<img width="507" alt="Screen Shot 2021-04-19 at 4 25 13 pm" src="https://user-images.githubusercontent.com/870154/115190780-e4d8a100-a12b-11eb-9073-f0be436a7445.png">

Now:

<img width="433" alt="Screen Shot 2021-04-19 at 4 27 26 pm" src="https://user-images.githubusercontent.com/870154/115190977-236e5b80-a12c-11eb-9b70-f0ee74616076.png">

sidvishnoi replied:
Why role=link?

marcoscaceres replied:
See first point in https://github.com/w3c/respec/issues/2825#issuecomment-811516805

sidvishnoi reviewed (commented):
Will review in detail when I get on computer. Looking good though.

sidvishnoi commented on tests/spec/core/dfn-spec.js:
```diff
@@ -21,6 +21,22 @@ describe("Core — Definitions", () => {
+    const doc = await makeRSDoc(ops);
+    const sec = doc.getElementById("dfns");
+    const dfns = sec.querySelectorAll("dfn");
+    expect(dfns.length).toHaveSize(4);
```
```suggestion
    expect(dfns).toHaveSize(4);
```

sidvishnoi commented on src/core/dfn-panel.runtime.js:
```diff
@@ -1,27 +1,38 @@
+    if (type === "keydown" && event.key !== "Enter") return;
+
+    const action = deriveAction(event);
+    const coords = deriveCoordinates(event);
```
We only need coordinates during "show", so lets move this call there.

marcoscaceres replied:
good point.

sidvishnoi commented on src/core/dfn-panel.runtime.js:
```diff
@@ -32,30 +43,50 @@ function dfnPanel() {
+    return { x: event.clientX, y: event.clientY };
+  }
+  // Otherwise, KeyboardEvent...
+  const target = /** @type HTMLElement */ (event.target);
```
Following should work fine also (cleaner IMO):

```suggestion
  /** @type HTMLElement */
  const target = event.target;
```

marcoscaceres replied:
Tried that originally, but doesn't seem to like it.

marcoscaceres replied:
<img width="1048" alt="Screen Shot 2021-04-19 at 1 25 00 pm" src="https://user-images.githubusercontent.com/870154/115177496-b1d5e380-a112-11eb-8e8e-505f29b46ebd.png">

sidvishnoi commented on src/core/dfn-panel.runtime.js:
```diff
@@ -32,30 +43,50 @@ function dfnPanel() {
-  if (clickTarget.closest("dfn, .index-term")) {
-    return hitALink ? null : "show";
+/**
+ * @param {Event} event
```
```suggestion
 * @param {MouseEvent|KeyboardEvent} event
```

marcoscaceres replied:
This one only deals with `.target`, which is only relevant for `Event`.

sidvishnoi commented on src/core/dfn-panel.runtime.js:
```diff
@@ -97,10 +128,90 @@ function displayPanel(dfn, panel, { x, y }) {
+      case "Tab": {
+        event.preventDefault();
+        let nextIndex = currentIndex + (event.shiftKey ? -1 : +1);
+        if (nextIndex < 0) {
+          nextIndex = lastIndex;
+        } else if (nextIndex > lastIndex) {
+          nextIndex = 0;
+        }
```
Hopefully not too clever:

```suggestion
        nextIndex = nextIndex - lastIndex * Math.floor(nextIndex / lastIndex);
```
(With `const lastIndex = anchors.length;`)

marcoscaceres replied:
I think that's bit too clever tbh (at least, I would find that to read/understand).

sidvishnoi commented on src/core/dfn-panel.runtime.js:
```diff
@@ -97,10 +128,90 @@ function displayPanel(dfn, panel, { x, y }) {
+ * @param {HTMLElement} dfn
+ * @returns
+ */
+function createTrapListener(anchors, panel, dfn) {
```
Maybe we can use the trap we've in `core/ui`:
https://github.com/w3c/respec/blob/77ab0c7507d9960c50745dde954a87e199e2be81/src/core/ui.js#L94-L120

marcoscaceres replied:
Turns out the ui.js code is actually busted in Safari :( (but works fine in Chrome!) 

In Safari, we need the explicit anchor list, otherwise Safari just moves the cursor to wherever it thinks it's best... sometime to the browser's address bar... otherwise, to other links.

marcoscaceres replied:
Filed https://github.com/w3c/respec/issues/3462

sidvishnoi commented on src/core/dfn-panel.runtime.js:
```diff
@@ -97,10 +128,90 @@ function displayPanel(dfn, panel, { x, y }) {
+      // Hitting "Escape" returns focus to dfn.
+      case "Escape":
+        hidePanel(panel);
+        dfn.focus();
+        break;
+    }
+    anchors.item(currentIndex).focus();
```
Otherwise it moves focus to current item again..

```suggestion
        dfn.focus();
        return;
    }
    anchors.item(currentIndex).focus();
```

sidvishnoi reviewed (changes requested):
Almost there...

jnurthen commented on src/core/dfn-panel.js:
```diff
@@ -24,6 +24,11 @@ export async function run() {
+    // allowing keyboard action as needed.
+    el.tabIndex = 0;
+    el.setAttribute("role", "link");
+    el.setAttribute("aria-haspopup", "dialog");
```
I like having role=link - this resolves my 1 big concern that this was adding widget semantics to a structural role. Links are widgets so it is natural to focus them. 
1 improvement would be to add an accessible description to indicate that this will "show usages of foo" - the easiest way to do this would be using the title attribute.

jnurthen commented on src/core/dfn-panel.runtime.js:
```diff
@@ -32,30 +43,50 @@ function dfnPanel() {
+ */
+function deriveCoordinates(event) {
+  if (event instanceof MouseEvent) {
+    return { x: event.clientX, y: event.clientY };
```
I did notice the dialog being in the incorrect position when running with NVDA. I think this is because NVDA uses top left of the bounding box as clientX and clientY when clicking. Not sure what we can do about it (and whether it is an issue that needs resolving)

marcoscaceres replied:
Yeah, I noticed that Chrome + VO sometimes does the same thing... but sometimes it makes the event's `.clientX = 0`. I'd be hesitant to fix that unless it becomes a problem, as it seems to be a browser bug.

jnurthen replied:
I'm wondering if we get a zero (highly unlikely) we simulate a click in the middle of the bounding box

marcoscaceres replied:
Chrome indeed sends clientX/clientY zeros... this kinda bad because it makes AT detectable. Anyway, added some code to handle that.

jnurthen reviewed (approved):
I'm happy to go forward as-is with the comments being addressed in subsequent issues if you want as this is a huge improvement over the current code.

sidvishnoi commented on src/core/dfn-finder.js:
```diff
@@ -182,7 +182,10 @@ export function decorateDfn(dfnElem, idlAst, parent, name) {
-    if (last === "") last = "the-empty-string";
+    if (last === "") {
+      last = "the-empty-string";
+      dfnElem.setAttribute("aria-label", "the empty string");
```
Nit: gotta reduce that bundle size 😄
```suggestion
      dfnElem.setAttribute("aria-label", last);
```

marcoscaceres replied:
"the-empty-string" is different to "the empty string", so will keep this one.

sidvishnoi replied:
Ah. Missed that.

sidvishnoi commented on src/core/dfn-panel.runtime.js:
```diff
@@ -97,10 +128,90 @@ function displayPanel(dfn, panel, { x, y }) {
+        dfn.focus();
+        return;
+    }
+    anchors.item(currentIndex).focus();
```
currentIndex is updated only on Tab, so maybe move this line there.

marcoscaceres commented on src/core/dfn-panel.js:
```diff
@@ -41,15 +41,28 @@ function createPanel(dfn) {
+        <a
+          class="self-link"
+          href="${href}"
+          aria-label="Permalink for definition: ${definition}"
```
Expanded description to say the permalink closes the dialog.

marcoscaceres commented:
Thanks again for your help @jnurthen! 

@sidvishnoi, should be all good to go 🤞

sidvishnoi commented on src/core/dfn-panel.runtime.js:
```diff
@@ -97,10 +137,86 @@ function displayPanel(dfn, panel, { x, y }) {
+      // Hitting "Tab" traps us in a nice loop around elements.
+      case "Tab": {
+        event.preventDefault();
+        let nextIndex = currentIndex + (event.shiftKey ? -1 : +1);
+        if (nextIndex < 0) {
+          nextIndex = lastIndex;
+        } else if (nextIndex > lastIndex) {
+          nextIndex = 0;
+        }
+        currentIndex = nextIndex;
```
```suggestion
        currentIndex += event.shiftKey ? -1 : +1;
        if (currentIndex < 0) {
          currentIndex = lastIndex;
        } else if (currentIndex > lastIndex) {
          currentIndex = 0;
        }
```

sidvishnoi commented on src/core/dfn-panel.runtime.js:
```diff
@@ -97,10 +137,86 @@ function displayPanel(dfn, panel, { x, y }) {
+
+      // Hitting "Enter" on an anchor releases the trap.
+      case "Enter":
+        hidePanel(panel);
```
The click behavior is to "dock" the panel on anchor click. Is this due to focus issues?

marcoscaceres replied:
just feels more natural to close and move the cursor... and yes, it's about focus too.

sidvishnoi replied:
Let's stick with hide for now. Docking will require to move the focus to permalink as well as add ability to jump back into panel, which complicates things.

sidvishnoi reviewed (approved):
Great improvement over the current version!

marcoscaceres commented:
Great work everyone! Bouncing around specs with keyboard is now quite nice. 

There are a few browser bugs:

  * Chrome seems to  .focus(), but doesn't visually show the focus.  
  * [Firefox fires two events](https://bugzilla.mozilla.org/show_bug.cgi?id=1695636) (keyup + click) - hi @saschanaz!! ❤️

but we will need to fix those in browsers.
